File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        
        List<List<String>> ans = new ArrayList<>();
        List<String> path = new ArrayList<>();
        
        dfs(0, s, path, ans);
        return ans;
    }
    
    private void dfs(int idx, String s, List<String> path, List<List<String>> ans){
        
        if(idx >= s.length()){
            ans.add(new ArrayList<>(path));
            return;
        }
        
        for(int i=idx; i<s.length(); i++){
            if(isPalindrome(s, idx, i)){
                path.add(s.substring(idx, i+1));
                dfs(i+1, s, path, ans);
                path.remove(path.size()-1);
            }
        }
    }
    
    boolean isPalindrome(String s, int start, int end){
        while(start <= end){
            if(s.charAt(start++) != s.charAt(end--)) return false;
        } 
        
        return true;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string& s, int l, int r){
        
        while(l<r){
            if(s[l] != s[r])
                return false;
            l++;
            r--;
        }
        
        return true;
    }
    
    void backtrack(string& s, int idx, vector<string> ds, vector<vector<string>>& result, int n){
        
        if(idx >= s.length()){
            result.push_back(ds);
            return;
        }
        
        for(int i=idx; i<n; i++){
            if(isPalindrome(s, idx, i)){
                ds.push_back(s.substr(idx, i-idx+1));
                backtrack(s, i+1, ds, result, n);
                ds.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        
        int n = s.length();
        
        vector<vector<string>> result;
        vector<string> ds;
        
        backtrack(s, 0, ds, result, n);
        return result;
    }
};